controllers/imageclusterinstall_controller.go

```diff
@@ -228,20 +232,19 @@ func (r *ImageClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
 	}
 
 	if ici.Status.BareMetalHostRef != nil && !v1alpha1.BMHRefsMatch(ici.Spec.BareMetalHostRef, ici.Status.BareMetalHostRef) {
-		if _, err := r.removeBMHImage(ctx, ici.Status.BareMetalHostRef); client.IgnoreNotFound(err) != nil {
+		if _, err := r.removeBMHDataImage(ctx, bmh, ici.Status.BareMetalHostRef); client.IgnoreNotFound(err) != nil {
 			log.WithError(err).Errorf("failed to remove image from BareMetalHost %s/%s", ici.Status.BareMetalHostRef.Namespace, ici.Status.BareMetalHostRef.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
 	if bmh != nil {
-		// in case image data was changed, bmh has image url configured and bmh is not provisioning yet,
-		// we should remove image from it in order to invalidate ironic cache
-		if bmh.Spec.Image != nil && bmh.Status.Provisioning.State != bmh_v1alpha1.StateProvisioned &&
-			bmh.Status.Provisioning.State != bmh_v1alpha1.StateProvisioning &&
-			updated {
-			r.Log.Info("Image data was changed, removing image from BareMetalHost")
-			removed, err := r.removeBMHImage(ctx, ici.Spec.BareMetalHostRef)
+		// In case data image was changed and bmh is not externally provisioned yet,
+		// we should remove the DataImage from it in order to invalidate ironic cache
+		dataImage := r.getBmhDataImage(ctx, bmh)
+		if dataImage != nil && bmh.Status.Provisioning.State != bmh_v1alpha1.StateExternallyProvisioned && updated {
+			r.Log.Info("Data image was changed, removing data image")
+			removed, err := r.removeBMHDataImage(ctx, bmh, ici.Spec.BareMetalHostRef)
```

Comment:
IBIO set the BMH provisioning state to externallyProvisioned right after we create the dataImage.
Do we need this code? 
I see it was here, it's just not clear what the scenario + we know that it doesn't actually work (the data image won't get deleted)

Reply:
Scenario is like this :
1. You set wrong machine cidr
2. Validation failed.
3. You change ici
4. Invalidate cache

Reply:
if validation fails you shouldn't create the data image so there's nothing to invalidate.

Is there a valid scenario for this?

Reply:
I believe we can remove, yeah missed that we set it after it. Though we need to create scenario that allows to update ici and clean the cache